R/MQR_rq.R

```diff
@@ -0,0 +1,124 @@
+#' }
+#' @param exclude_train A column name in \code{data} indicating if a row should be excluded from model
+#' training, i.e. if it contains bad data (will be coerced to \code{logical}). Alternatively,
+#' an \code{integer} or \code{logical} vector with length equal to the number of rows in \code{data} indicating
+#' the same. Rows labelled \code{TRUE} are excluded from model training.
+#' @param sort \code{boolean} Sort quantiles using \code{SortQuantiles()}?
+#' @param sort_limits \code{Limits} argument to be passed to \code{SortQuantiles()}. Constrains quantiles to upper and lower limits given by \code{list(U=upperlim,L=lowerlim)}.
+#' @param ... Additional arguments passed to \code{rq}.
+#' @details The returned predictive quantiles are those produced out-of-sample for each
+#' cross-validation fold (using models trained on the remaining folds but not "Test" data).
+#' Predictive quantiles corresponding to "Test" data are produced using models trained on all
+#' non-test data.
+#' @return Returns a \code{list} containing predictive quantiles (in a \code{MultiQR} object) and \code{rq} models.
+#' @keywords Quantile Regression
+#' @export
+qreg_mrq <- function(data,
+                     formula,
+                     quantiles=c(0.25,0.5,0.75),
+                     offset = NULL,
+                     cv_folds=NULL,
+                     exclude_train=NULL,
+                     sort=T,sort_limits=NULL,
+                     ...){
+  
+  
+  # Set-up cv folds & do checks
+  cv_labs <- cv_control(data = data,cv_folds = cv_folds)
+  
+  # Exclude points from training? & do checks
+  exclude_idx <- exclude_fun(data = data,exclude_train = exclude_train)
+  
+  # Initialise output container
+  FINAL_OUTPUT <- list(mqr_pred=data.table(matrix(as.numeric(NA),
+                                                  ncol = length(quantiles),
+                                                  nrow = nrow(data))),
+                       call=match.call(),
+                       kfold_index = cv_labs$idx,
+                       model_names = cv_labs$fold_loop,
+                       exclude_index = exclude_idx,
+                       models = list(),
+                       sorted = list(sort=sort,
+                                     sort_limits=sort_limits),
+                       default_model=if("Test"%in%cv_labs$fold_loop){"Test"}else{cv_labs$fold_loop[1]})
+  colnames(FINAL_OUTPUT$mqr_pred) <- paste0("q",100*quantiles)
+  class(FINAL_OUTPUT) <- c("qreg_mrq",class(FINAL_OUTPUT))
+  
+  # Adjust formula if offset being used (rq does not handle offsets automatically in the same way as lm)
+  if(!is.null(offset)){
+    formula = as.formula(paste0(c(terms(formula)[[2]],"-",offset," ~ ",terms(formula)[[3]]),collapse = ""))
+  }
+  
+  for(fold in FINAL_OUTPUT$model_names){
+    
+    print(paste0("Multiple rq, kfold=",fold))
+    
+    for(fold in FINAL_OUTPUT$model_names){
```

Comment:
Could you explain why you are looping over folds twice, but only using the fold name from the second loop?  Maybe we can add something to the documentation to explain, otherwise the output when you run the function looks a little confusing? 

[1] "Multiple rq, kfold=Fold 1"
[1] "MQR, kfold=Fold 1"
[1] "MQR, kfold=Fold 2"
[1] "MQR, kfold=Fold 3"
[1] "MQR, kfold=Test"
[1] "Multiple rq, kfold=Fold 2"
[1] "MQR, kfold=Fold 1"
[1] "MQR, kfold=Fold 2"
[1] "MQR, kfold=Fold 3"
[1] "MQR, kfold=Test"
[1] "Multiple rq, kfold=Fold 3"
[1] "MQR, kfold=Fold 1"
[1] "MQR, kfold=Fold 2"
[1] "MQR, kfold=Fold 3"
[1] "MQR, kfold=Test"
[1] "Multiple rq, kfold=Test"
[1] "MQR, kfold=Fold 1"
[1] "MQR, kfold=Fold 2"
[1] "MQR, kfold=Fold 3"
[1] "MQR, kfold=Test"

Reply:
Thanks! It's kind of you to ask me to explain. This is a silly error, there should only be the one loop.